situp-plugins/peer-forwarder/src/main/java/com/amazon/situp/plugins/processor/peerforwarder/PeerForwarder.java

```diff
@@ -0,0 +1,166 @@
+        // TODO: replace hardcoded port with customization
+        return Clients.builder(String.format("gproto+http://%s:21890/", ipAddress))
+                .writeTimeout(Duration.ofSeconds(peerForwarderConfig.getTimeOut()))
+                .build(TraceServiceGrpc.TraceServiceBlockingStub.class);
+    }
+
+    @Override
+    public List<Record<ExportTraceServiceRequest>> execute(final Collection<Record<ExportTraceServiceRequest>> records) {
+        final Map<String, List<ResourceSpans>> peerGroupedRS = new HashMap<>();
+        for (final String peerIp: peerIps) {
+            peerGroupedRS.put(peerIp, new ArrayList<>());
+        }
+        for (final Record<ExportTraceServiceRequest> record: records) {
+            for (final ResourceSpans rs: record.getData().getResourceSpansList()) {
+                final List<Map.Entry<String, ResourceSpans>> rsBatch = splitByTrace(rs);
+                for (final Map.Entry<String, ResourceSpans> entry: rsBatch) {
+                    final String traceId = entry.getKey();
+                    final ResourceSpans newRS = entry.getValue();
+                    final String peerIp = getHostByConsistentHashing(traceId);
+                    peerGroupedRS.get(peerIp).add(newRS);
+                }
+            }
+        }
+
+        // Send ExportTraceServiceRequest to designated peer
+        final List<Record<ExportTraceServiceRequest>> results = new ArrayList<>();
+        for (final String peerIp: peerIps) {
+            final TraceServiceGrpc.TraceServiceBlockingStub client = peerClients.getOrDefault(peerIp, null);
+            ExportTraceServiceRequest.Builder currRequestBuilder = ExportTraceServiceRequest.newBuilder();
+            int currSpansCount = 0;
+            for (final ResourceSpans rs: peerGroupedRS.get(peerIp)) {
+                final int rsSize = getResourceSpansSize(rs);
+                if (currSpansCount >= peerForwarderConfig.getMaxNumSpansPerRequest()) {
+                    final ExportTraceServiceRequest currRequest = currRequestBuilder.build();
+                    processRequest(client, currRequest, results);
+                    currRequestBuilder = ExportTraceServiceRequest.newBuilder();
+                    currSpansCount = 0;
+                }
+                currRequestBuilder.addResourceSpans(rs);
+                currSpansCount += rsSize;
+            }
+            // last request
+            final ExportTraceServiceRequest currRequest = currRequestBuilder.build();
+            processRequest(client, currRequest, results);
+        }
+        return results;
+    }
+
+    private void processRequest(final TraceServiceGrpc.TraceServiceBlockingStub client,
+                                final ExportTraceServiceRequest request,
+                                final List<Record<ExportTraceServiceRequest>> results) {
+        if (client != null) {
+            try {
+                client.export(request);
```

Comment:
Yeah. I think we should guard against local ip appearing in the peers to prevent this from happening.

Reply:
Each host needs to have a complete list of peers (including itself) so that hashing is consistent regardless of the host processing the request. 

If each host's self-address is removed from the peers list, then all hosts will have inconsistent lists. Here's a simple two-host scenario where the "remove self" logic doesn't quite work out.

Two hosts A and B:
* host list in config file: `192.168.1.hostA`, `192.168.1.hostB`
* host list in hostA (after removing itself): `192.168.1.hostB` 
* host list in hostB (after removing itself): `192.168.1.hostA`

In this case, calling `getHostByConsistentHashing` on hostA will always point to hostB, and calling on hostB will always point to hostA resulting in endless requests bouncing between the two. We might even want to rename "peerIps" to something else because we expect the list to contain addresses of ALL hosts, not just the peers of the current hosts.

---
Ultimately, the logic needs to hash on trace ID to determine the correct host to process this request, then:
  * If the correct host IS the current host, don't forward the request and instead process it locally
  * If the correct host is elsewhere on the network, forward the request and don't process locally

This exit-condition to stop forwarding is what's missing in the current implementation.

Reply:
> Each host needs to have a complete list of peers (including itself) so that hashing is consistent regardless of the host processing the request.

@wrijeff I did keep a complete list of peers although it is not in the input list. If you look at the constructor, the following procedure is done

1. removes public Ip of local host if it appears in the input list
2. create client for each of the remote ips (not including localhost)
3. add public ip of localhost into the list
4. sort the list to maintain the consistent order across SITUPs (this might become unnecessary once hash ring is introduced)

```
public PeerForwarder(final PluginSetting pluginSetting) {
        peerForwarderConfig = PeerForwarderConfig.buildConfig(pluginSetting);
        peerIps = new ArrayList<>(new HashSet<>(peerForwarderConfig.getPeerIps()));
        final String localPublicIp;
        try {
            localPublicIp = getLocalPublicIp();
            peerIps.remove(localPublicIp);
        } catch (IOException e) {
            throw new RuntimeException("Cannot get localhost public IP.", e);
        }
        peerClients = peerIps.stream().collect(Collectors.toMap(ip-> ip, ip-> createGRPCClient(ip)));
        peerIps.add(localPublicIp);
        Collections.sort(peerIps);
    }
```

So the above scenario should not happen.

Reply:
Ah thanks, I missed the line adding the IP back into the list. I still think this should be solved without relying on an external service to provide the public IP - would be interested in how this approach would have to change.